src/python/pants/engine/internals/graph.py

```diff
@@ -98,85 +98,88 @@ async def resolve_unexpanded_targets(addresses: Addresses) -> UnexpandedTargets:
-        return Subtargets(build_target, ())
-
-    # Generate a subtarget per source.
-    paths = await Get(SourcesPaths, SourcesPathsRequest(build_target[Sources]))
-    wrapped_subtargets = await MultiGet(
-        Get(
-            WrappedTarget,
-            Address,
-            generate_subtarget_address(address, full_file_name=subtarget_file),
-        )
-        for subtarget_file in paths.files
+def target_types_to_generate_targets_requests(
+    union_membership: UnionMembership,
+) -> TargetTypesToGenerateTargetsRequests:
+    return TargetTypesToGenerateTargetsRequests(
+        {
+            request_cls.target_class: request_cls  # type: ignore[misc]
+            for request_cls in union_membership.get(GenerateTargetsRequest)
+        }
     )
-    return Subtargets(build_target, tuple(wt.target for wt in wrapped_subtargets))
 
 
 @rule
 async def resolve_target(
     address: Address,
     registered_target_types: RegisteredTargetTypes,
     union_membership: UnionMembership,
+    target_types_to_generate_requests: TargetTypesToGenerateTargetsRequests,
 ) -> WrappedTarget:
     if address.is_file_target:
-        build_target = await Get(WrappedTarget, Address, address.maybe_convert_to_build_target())
-        subtarget = generate_subtarget(
-            build_target.target, full_file_name=address.filename, union_membership=union_membership
+        wrapped_generator_tgt = await Get(
+            WrappedTarget, Address, address.maybe_convert_to_build_target()
         )
-        return WrappedTarget(subtarget)
+        generator_tgt = wrapped_generator_tgt.target
+        # Not all languages support file targets (Go and Terraform). If we got a file address, use
+        # that as an alias for the BUILD target.
+        #
+        # This code path will go away once we remove file addresses in favor of generated target
+        # addresses, e.g. path/to:generator:generated, because we will error in that case when
+        # there is no generator for the address.
+        if not target_types_to_generate_requests.is_generator(generator_tgt):
+            return wrapped_generator_tgt
+        generate_request = target_types_to_generate_requests[type(generator_tgt)]
+        generated = await Get(
+            GeneratedTargets, GenerateTargetsRequest, generate_request(generator_tgt)
+        )
+        requested = next((tgt for tgt in generated if tgt.address == address), None)
+        if requested is None:
+            bulleted_list_sep = "\n  * "
+            generated_str = bulleted_list_sep.join(tgt.address.spec for tgt in generated)
+            raise ValueError(
+                f"The address `{address}` is not generated by the `{generator_tgt.alias}` target "
+                f"`{generator_tgt.address}`, which only generates these addresses:\n"
+                f"{bulleted_list_sep}{generated_str}\n\n"
+                "Did you mean to use one of those addresses?"
```

Comment:
New error message. The suggestion could probably be improved...I got tired of staring at this code all day and will probably revisit this soon.

Example:

```
❯ ./pants filedeps src/python/pants/util/strutil_test.py:util
ValueError: The address `src/python/pants/util/strutil_test.py` is not generated by the `python_library` target `src/python/pants/util`, which only generates these addresses:

  * src/python/pants/util/__init__.py
  * src/python/pants/util/collections.py
  * src/python/pants/util/contextutil.py
  * src/python/pants/util/dirutil.py
  * src/python/pants/util/docutil.py
  * src/python/pants/util/enums.py
  * src/python/pants/util/eval.py
  * src/python/pants/util/filtering.py
  * src/python/pants/util/frozendict.py
  * src/python/pants/util/logging.py
  * src/python/pants/util/memo.py
  * src/python/pants/util/meta.py
  * src/python/pants/util/ordered_set.py
  * src/python/pants/util/osutil.py
  * src/python/pants/util/rwbuf.py
  * src/python/pants/util/socket.py
  * src/python/pants/util/strutil.py

Did you mean to use one of those addresses?
```

Nb that this will apply soon when you, for example, depend on an incorrect target generated from `python_requirements`! E.g. `//:reqs::Djangooo`